### PR TITLE
Public Type System and Type Annotations [OLD]

### PR DESCRIPTION
**Pull request by boggle**

### CODE REVIEW

**Comment by thobe** [comment]
The original branch for this PR seems to have been removed. That seems premature since the PR is still open. It means that for example viewing the whole file does not work.

**Comment by systay** [comment]
Stuff broke slightly when moving repositories. You can view it here: https://github.com/systay/cypher/blob/4eaded0fb977c05149f4a3d39dd582468662ef71/cip/CIP2015-09-16-public-type-system-type-annotation.adoc

File: cip/CIP2015-09-16-public-type-system-type-annotation.adoc [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+== Abstract
+
+This CIP specifies the public part of the type system of Cypher and introduces literal syntax for writing down types as well as new syntax for type annotations. The proposed type annotation syntax does not change the Cypher grammar right now but rather is intended as a template to be used by future CIPs. This CIP also informatively discusses future syntax that might use the introduced literal type syntax.
+
+== Introduction
+=== Background and Motivation
+In order to support currently ongoing work for allowing users to define _procedures_ and _functions_ that interact with Cypher, it has become necessary to specify types explicitly, especially as part of procedure and function signatures. It is thus necessary to at least introduce syntax for all valid, supported public types as well as syntax for type ascription.
+
+=== Overview
+For the purposes of this CIP, a type system is a system of classifying expressions according to what kinds of values they may evaluate to during the execution of a Cypher query. It does *not *include a specification of a type inferencing and checking algorithm for Cypher queries.
```
Comment (systay) [inline cip/CIP2015-09-16-public-type-system-type-annotation.adoc]: Is it only about what types expression results will have? Isn't it also about expectations of value types?

File: cip/CIP2015-09-16-public-type-system-type-annotation.adoc [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+ ** Lists (ordered sequences with random access) of a given type T
+ ** `LIST<T1>`` is a subtype of `LIST<T2>`` if `T1` is a subtype of `T2`. This works since values in Cypher are immutable. Adding an element of type `T` to a `LIST<S>`` would produce a new list of type `LIST<R>``, where `R` is the nearest common supertype of `T` and `S`
+ * `NODE`
+ ** Nodes are maps that have labels and relationships
+ * `STRING`
+ ** Unicode Strings, i.e. `"Cypher"`, and `‘text’`
+ * `RELATIONSHIP`
+ ** Relationships are maps that have a type and end points
+ * `PATH`
+ ** List of alternating nodes and relationships
```
Comment (thobe) [inline cip/CIP2015-09-16-public-type-system-type-annotation.adoc]: This is a REALLY BAD way of interacting with, viewing, or defining a path. Let's stop repeating this mistake.

File: cip/CIP2015-09-16-public-type-system-type-annotation.adoc [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+ * `IDENTITY`
+ ** Identities are considered part of the current type system (since the +++<u>Identity type CIP</u>+++ has been accepted), even though it has not been implemented at the time of writing this CIP
+ * `MAP`
+ ** Maps to values of any type, i.e. { name: "Svensson" }. Note that keys are identifiers, not strings.
+ ** This is also the supertype for `NODE` and `RELATIONSHIP`
+ * `LIST<T>`
+ ** Lists (ordered sequences with random access) of a given type T
+ ** `LIST<T1>`` is a subtype of `LIST<T2>`` if `T1` is a subtype of `T2`. This works since values in Cypher are immutable. Adding an element of type `T` to a `LIST<S>`` would produce a new list of type `LIST<R>``, where `R` is the nearest common supertype of `T` and `S`
+ * `NODE`
+ ** Nodes are maps that have labels and relationships
```
Comment (thobe) [inline cip/CIP2015-09-16-public-type-system-type-annotation.adoc]: "Nodes are maps..." feels like a really strange way of defining what a node is.

File: cip/CIP2015-09-16-public-type-system-type-annotation.adoc (hunk 2 above)
Comment (systay) [inline cip/CIP2015-09-16-public-type-system-type-annotation.adoc]: What is a better definition?

File: cip/CIP2015-09-16-public-type-system-type-annotation.adoc [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+ * `STRING`
+ ** Unicode Strings, i.e. `"Cypher"`, and `‘text’`
+ * `RELATIONSHIP`
+ ** Relationships are maps that have a type and end points
```
Comment (thobe) [inline cip/CIP2015-09-16-public-type-system-type-annotation.adoc]: I have the same irk here as I do with the definition of node. Actually here the definition of what a relationship is feels even stranger than what the prior definition of node did.

File: cip/CIP2015-09-16-public-type-system-type-annotation.adoc [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+== Introduction
+=== Background and Motivation
+In order to support currently ongoing work for allowing users to define _procedures_ and _functions_ that interact with Cypher, it has become necessary to specify types explicitly, especially as part of procedure and function signatures. It is thus necessary to at least introduce syntax for all valid, supported public types as well as syntax for type ascription.
+
+=== Overview
+For the purposes of this CIP, a type system is a system of classifying expressions according to what kinds of values they may evaluate to during the execution of a Cypher query. It does *not *include a specification of a type inferencing and checking algorithm for Cypher queries.
+
+Cypher today has a rich type system that unifies Java implementation types into a coherent type hierarchy and an accompanying type checker that supports rich types with generics and type unions. So far it has not been possible to explicitly specify the type of an expression in a query.
+
+This CIP proposes to introduce syntax for type literals and for type annotations.
```
Comment (systay) [inline cip/CIP2015-09-16-public-type-system-type-annotation.adoc]: "type annotations" is mentioned but not defined a couple of times in this document

File: cip/CIP2015-09-16-public-type-system-type-annotation.adoc [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+ * `PATH`
+ ** List of alternating nodes and relationships
+ * `NUMBER`
+ ** Supertype to both `INTEGER` and `FLOAT`
+ * `DATETIME`
+ ** An instant capturing the date, the time, and the timezone. More information about the various temporal types, see the _CIP20150806 - Date and Time_
+ * `LOCALDATETIME`
+ ** An instant capturing the date and the time, but not the time zone.
+ * `DATE`
+ ** An instant capturing the date, but not the time, nor the time zone.
```
Comment (thobe) [inline cip/CIP2015-09-16-public-type-system-type-annotation.adoc]: `DATE` has the same definition as `LOCALDATE`, and `TIME` is missing...

File: cip/CIP2015-09-16-public-type-system-type-annotation.adoc [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+
+=== Alternatives
+As part of writing this CIP, many syntax alternatives have been considered:
+
+* Alternative name for the `STRING` type: `TEXT`, `UNICODE`, `LIST<CHAR>`
+* Alternative name for the `LIST` type: `ARRAY`, `COLLECTION`, `SEQUENCE`, `VECTOR`
+* Alternative syntax for type parameters: `LIST OF T`, `LIST[T]`, `LIST<T>`, `LIST T`
+* Alternative syntax for type annotations: `(STRING) expr`, `STRING expr`, `expr: `T`
+
+== What others do
```
Comment (systay) [inline cip/CIP2015-09-16-public-type-system-type-annotation.adoc]: Like someone else has commented somewhere else - this section only deals with casting and coercing, and not around the type system per se. A discussion on the type system compared to our competitors would be interesting

File: cip/CIP2015-09-16-public-type-system-type-annotation.adoc (hunk 2 above)
Comment (petraselmer) [inline cip/CIP2015-09-16-public-type-system-type-annotation.adoc]: Definition of a path (in which the classical definition has been abused somewhat): A path from a node `n1` to a node `ni` - corresponding to a walk in the graph from `n1` to `ni` - is a sequence n1,r1,n2,r2,...,r(i-1),ni of alternating nodes and relationships such that for 1 <= j <= i, rj is incident with nj and n(j+1).

By the way, I didn't go into precisely how these are represented...

**Comment by boggle** [comment]
Superseded by https://github.com/opencypher/openCypher/pull/12. Closing.

**Pull request closed without merging**